src/Pim/Bundle/DataGridBundle/Datasource/AssociatedProductDatasource.php

```diff
@@ -0,0 +1,316 @@
+
+        if (Directions::ASCENDING === $this->sortOrder) {
+            $rows['data'] = $this->getProductsSortedByIsAssociatedAsc(
+                $associatedProductsIdentifiers,
+                $limit,
+                $searchAfter,
+                $locale,
+                $scope
+            );
+        } elseif (Directions::DESCENDING === $this->sortOrder) {
+            $rows['data'] = $this->getProductsSortedByIsAssociatedDesc(
+                $associatedProductsIdentifiers,
+                $limit,
+                $searchAfter,
+                $locale,
+                $scope
+            );
+        } else {
+            $rows['data'] = $this->normalizeProducts($productCursor, $associatedProductsIdentifiers, $locale, $scope);
+        }
+
+        return $rows;
+    }
+
+    /**
+     * @param ProductInterface $product
+     * @param string           $associationTypeId
+     *
+     * @return string[]
+     */
+    protected function getAssociatedProductsIdentifiers(ProductInterface $product, $associationTypeId)
+    {
+        $identifiers = [];
+
+        foreach ($product->getAssociations() as $association) {
+            if ($association->getAssociationType()->getId() === (int) $associationTypeId) {
+                foreach ($association->getProducts() as $associatedProduct) {
+                    $identifiers[] = $associatedProduct->getIdentifier();
+                }
+            }
+        }
+
+        return $identifiers;
+    }
+
+    /**
+     * @param array  $associatedProductsIdentifiers
+     * @param string $limit
+     * @param string $searchAfter
+     * @param string $locale
+     * @param string $scope
+     *
+     * @return array
+     */
+    protected function getProductsSortedByIsAssociatedDesc(
+        array $associatedProductsIdentifiers,
+        $limit,
+        $searchAfter,
+        $locale,
+        $scope
+    ) {
+        $associatedPdts = $this->getAssociatedProducts(
+            $associatedProductsIdentifiers,
+            $limit,
+            $searchAfter,
+            $locale,
+            $scope
+        );
+
+        $nonAssociatedPdts = [];
+        $nbAssociated = count($associatedPdts);
+        if ($limit > $nbAssociated) {
+            $limit -= $nbAssociated;
+            $searchAfter -= count($associatedProductsIdentifiers);
+            $searchAfter = max($searchAfter, 0);
+
+            $nonAssociatedPdts = $this->getNonAssociatedProducts(
+                $associatedProductsIdentifiers,
+                $limit,
+                $searchAfter,
+                $locale,
+                $scope
+            );
+        }
+
+        return $associatedPdts + $nonAssociatedPdts;
+    }
+
+    /**
+     * @param array  $associatedProductsIdentifiers
+     * @param string $limit
+     * @param string $searchAfter
+     * @param string $locale
+     * @param string $scope
+     *
+     * @return array
+     */
+    protected function getProductsSortedByIsAssociatedAsc(
```

Comment:
Explain here that the non associated products are put at the begining of the list